{
  assertEquals("state:invalid",TTLObject.State.INVALID,object.getState());
  assertEquals("no dones",0,doneCount.get());
  object.refresh();
  assertEquals("state:generate",TTLObject.State.GENERATE,object.getState());
  assertEquals("no dones",0,doneCount.get());
  object.refresh();
  assertEquals("state:generate",TTLObject.State.GENERATE,object.getState());
  assertEquals("no dones",0,doneCount.get());
  object.getObject();
  assertEquals("state:valid",TTLObject.State.VALID,object.getState());
  assertEquals("one done",1,doneCount.get());
  object.getObject();
  assertEquals("state:valid",TTLObject.State.VALID,object.getState());
  assertEquals("one done",1,doneCount.get());
  object.getObject();
  assertEquals("state:valid",TTLObject.State.VALID,object.getState());
  assertEquals("one done",1,doneCount.get());
  object.getObject();
  assertEquals("state:valid",TTLObject.State.VALID,object.getState());
  assertEquals("one done",1,doneCount.get());
  object.refresh();
  assertEquals("state:generate",TTLObject.State.GENERATE,object.getState());
  assertEquals("one done",1,doneCount.get());
  object.getObject();
  assertEquals("state:valid",TTLObject.State.VALID,object.getState());
  assertEquals("two dones",2,doneCount.get());
  object.refresh();
  assertEquals("state:generate",TTLObject.State.GENERATE,object.getState());
  assertEquals("two dones",2,doneCount.get());
  throwException.set(new Thrower(){
    public void throwException() throws Exception {
      throw new Exception("exc1");
    }
  }
);
  TTLObject.ObjectException caught=null;
  try {
    object.getObject();
  }
 catch (  TTLObject.ObjectException e) {
    caught=e;
  }
 finally {
    assertNotNull("exception thrown",caught);
    assertEquals("correct exception thrown","exc1",caught.getCause().getMessage());
  }
  assertEquals("two dones",3,doneCount.get());
  caught=null;
  try {
    object.getObject();
  }
 catch (  TTLObject.ObjectException e) {
    caught=e;
  }
 finally {
    assertNotNull("exception thrown",caught);
    assertEquals("correct exception rethrown","exc1",caught.getCause().getMessage());
  }
  object.refresh();
  assertEquals("two dones",3,doneCount.get());
  object.getObject();
  assertEquals("two dones",4,doneCount.get());
  object.set("one");
  assertEquals("one","one",object.getObject());
  assertEquals("two dones",4,doneCount.get());
  object.set("two");
  object.refresh();
  assertEquals("two",(String)null,object.getObject());
  assertEquals("two dones",5,doneCount.get());
}
